config/routes.rb

```diff
@@ -19,6 +19,10 @@
     namespace :internal do
       get '/sessions' => 'sessions#show'
       put '/sessions' => 'sessions#update'
+
+      namespace :two_factor_authentication do
+        resource :webauthn, path: '/webauthn/:id', controller: :webauthn, only: [:update, :destroy]
```

Comment:
The two main reasons I opted for `resource` were:

- In aliasing the route, it felt strange to reference an alias defined on a `PUT` route (in this example) for [a reference used by the `DELETE` button](https://github.com/18F/identity-idp/blob/aduth-webauthn-manage-controller/app/views/users/webauthn/edit.html.erb#L27)
- Avoiding repetition of configuration, though maybe this is only `controller` now that I think of it

I'll poke around at it and see if it wouldn't be much trouble to update. That could also help avoid `path_names: { edit: '' }` which I felt was a little non-obvious in its purpose.

Reply:
> * In aliasing the route, it felt strange to reference an alias defined on a `PUT` route (in this example) for [a reference used by the `DELETE` button](https://github.com/18F/identity-idp/blob/aduth-webauthn-manage-controller/app/views/users/webauthn/edit.html.erb#L27)

We do this a lot for existing routes already, here's a quick grab of a few from `bundle exec rails routes`
```
                   api_internal_sessions GET             /api/internal/sessions(.:format)                                     api/internal/sessions#show
                                         PUT             /api/internal/sessions(.:format)                                     api/internal/sessions#update

                           user_password PATCH           (/:locale)/users/password(.:format)                                  users/reset_passwords#update {:locale=>/en|es|fr/}
                                         PUT             (/:locale)/users/password(.:format)                                  users/reset_passwords#update {:locale=>/en|es|fr/}
                                         POST            (/:locale)/users/password(.:format)                                  users/reset_passwords#create {:locale=>/en|es|fr/}

                 service_provider_revoke GET             (/:locale)/account/service_providers/:sp_id/revoke(.:format)         users/service_provider_revoke#show {:locale=>/en|es|fr/}
                                         DELETE          (/:locale)/account/service_providers/:sp_id/revoke(.:format)         users/service_provider_revoke#destroy {:locale=>/en|es|fr/}

                   account_reset_request GET             (/:locale)/account_reset/request(.:format)                           account_reset/request#show {:locale=>/en|es|fr/}
                                         POST            (/:locale)/account_reset/request(.:format)                           account_reset/request#create {:locale=>/en|es|fr/}

                    account_reset_cancel GET             (/:locale)/account_reset/cancel(.:format)                            account_reset/cancel#show {:locale=>/en|es|fr/}
                                         POST            (/:locale)/account_reset/cancel(.:format)                            account_reset/cancel#create {:locale=>/en|es|fr/}
````

Reply:
Trying to implement this, I'm bumping into what I suspect was partly what led me to `resource`, which is that the individual methods use a more verbose auto-generated name, even with the `:as` aliasing.

This code:

```rb
  namespace :api do
    namespace :internal do
      get '/sessions' => 'sessions#show'
      put '/sessions' => 'sessions#update'

      namespace :two_factor_authentication do
        put '/webauthn/:id' => 'webauthn#update', as: :webauthn
```

..gives: `api_internal_two_factor_authentication_webauthn_path`

vs.:

https://github.com/18F/identity-idp/blob/90f6ced7de5312b408bcf5d98a91277ea5118225/app/views/users/webauthn/edit.html.erb#L10

I might just be doing something wrong here, so I'll keep poking at it

Reply:
maybe it's time we just switch to `resources`, I like that we use the `only: [...]` so we don't generate extra unused routes

Reply:
> We do this a lot for existing routes already, here's a quick grab of a few from `bundle exec rails routes`

I think the key here was using `as: nil` on all but the first route, similar to what's done in the `user_password` case:

https://github.com/18F/identity-idp/blob/1eacd8af80c78c0a0b8f2f05562a742f400aab9c/config/routes.rb#L71-L73

That wasn't very obvious to me, and trying to reuse the same name resulted in an error instead:

```
raise ArgumentError, "Invalid route name, already in use: '#{name}' \n" \
```

Reply:
Ok, happy to leave as-is.

Sharing since I'd done the work already, but I believe this would be equivalent:

```diff
diff --git a/config/routes.rb b/config/routes.rb
index 250232077..78ab0a170 100644
--- a/config/routes.rb
+++ b/config/routes.rb
@@ -23,3 +23,4 @@ Rails.application.routes.draw do
       namespace :two_factor_authentication do
-        resource :webauthn, path: '/webauthn/:id', controller: :webauthn, only: [:update, :destroy]
+        put '/webauthn/:id' => 'webauthn#update', as: :webauthn
+        delete '/webauthn/:id' => 'webauthn#destroy', as: nil
       end
@@ -253,7 +254,5 @@ Rails.application.routes.draw do
     post '/manage/personal_key' => 'users/personal_keys#update'
-    resource :webauthn,
-             path: '/manage/webauthn/:id',
-             controller: 'users/webauthn',
-             path_names: { edit: '' },
-             only: [:edit, :update, :destroy]
+    get '/manage/webauthn/:id' => 'users/webauthn#edit', as: :edit_webauthn
+    put '/manage/webauthn/:id' => 'users/webauthn#update', as: :webauthn
+    delete '/manage/webauthn/:id' => 'users/webauthn#destroy', as: nil
```

Aside form `:webauthn` vs. `nil`,  I'm not sure how I feel on the naming of `edit_webauthn_path`. This aligns to [how Rails default URL helpers work](https://guides.rubyonrails.org/routing.html#path-and-url-helpers), but seemingly the reason that exists is because it's a different path pattern (`/:id/edit`) that we don't use.

Reply:
seeing this diff, I do think the explicit versions are clearer to read than the `resources` (especially the `/manage/webauthn` one), but ya if we think `resources` is clearer, I'd support that

Reply:
I'm happy with explicit routes.  Updated in c9bd16751.